scipy/linalg/flapack_other.pyf.src

```diff
@@ -47,64 +47,144 @@ subroutine <prefix2>tgsen(ijob,wantq,wantz,select,n,a,lda,b,ldb,alphar,alphai,be
-    integer intent(hide),depend(b) :: ldb=shape(b,0)
-    <ftype2> intent(out),dimension(n) :: alphar
-    <ftype2> intent(out),dimension(n) :: alphai
-    <ftype2> intent(out),dimension(n) :: beta
-    <ftype2> intent(in,out,copy),dimension(ldq,n) :: q
-    integer intent(hide),depend(q) :: ldq=shape(q,0)
-    <ftype2> intent(in,out,copy),dimension(ldz,n) :: z
-    integer intent(hide),depend(z) :: ldz=shape(z,0)
+    integer optional, intent(in),check(ijob>=0&&ijob<=5) :: ijob=4
+    logical optional, intent(in),check(wantq==0||wantq==1) :: wantq=1
+    logical optional, intent(in),check(wantz==0||wantz==1) :: wantz=1
+    logical intent(in),dimension(n),depend(n) :: select
+    integer intent(hide),depend(a) :: n=shape(a,0)
+    <ftype2> intent(in,out,copy,out=as),dimension(n,n) :: a
+    integer intent(hide),depend(a) :: lda=MAX(1,shape(a,0))
+    <ftype2> intent(in,out,copy,out=bs),dimension(n,n) :: b
+    integer intent(hide),depend(b) :: ldb = MAX(1, shape(b,0))
+    <ftype2> intent(out),dimension(n),depend(n) :: alphar
+    <ftype2> intent(out),dimension(n),depend(n) :: alphai
+    <ftype2> intent(out),dimension(n),depend(n) :: beta
+    <ftype2> intent(in,out,copy,out=qs),dimension(n,n),depend(n) :: q
+    integer intent(hide),depend(q) :: ldq=MAX(1,shape(q,0))
+    <ftype2> intent(in,out,copy,out=zs),dimension(n,n),depend(n) :: z
+    integer intent(hide),depend(z) :: ldz=MAX(1,shape(z,0))
     integer intent(out) :: m
     <ftype2> intent(out) :: pl
     <ftype2> intent(out) :: pr
     <ftype2> intent(out),dimension(2) :: dif
-    <ftype2> intent(out),dimension(MAX(lwork,1)) :: work
-    integer optional,intent(in),depend(m,n),check(lwork == -1 || lwork >= MAX(4*n+16,2*m*(n-m))) :: lwork=max(MAX(4*n+16,2*m*(n-m)),1)
-    integer intent(out),dimension(MAX(1,liwork)) :: iwork
-    integer optional,intent(in),depend(n),check(liwork == -1 || liwork >= n+6) :: liwork=n+6
+    <ftype2> intent(hide),dimension(MAX(lwork,1)) :: work
+    ! these lwork and liwork values are bare minimum estiamates only for cases ijob == 1,2,4
+    ! a separate lwork query is a prerequisite due to m dependence.
+    ! Depending on the m value lwork can go upto n**2 / 4 for n = 2m hence it is not
+    ! possible to give a minimal value without a potential excessive memory waste
+    integer optional,intent(in),depend(n),check(lwork == -1 || lwork >= 1) :: lwork=4*n+16
+    integer intent(hide),dimension(MAX(1,liwork)) :: iwork
+    integer optional,intent(in),depend(n),check(liwork == -1 || liwork >= 1) :: liwork=n+6
     integer intent(out) :: info
 
 end subroutine <prefix2>tgsen
 
+subroutine <prefix2>tgsen_lwork(ijob,wantq,wantz,select,n,a,lda,b,ldb,alphar,alphai,beta,q,ldq,z,ldz,m,pl,pr,dif,work,lwork,iwork,liwork,info)
+
+    fortranname <prefix2>tgsen
+    callstatement (*f2py_func)(&ijob,&wantq,&wantz,select,&n,a,&lda,&b,&ldb,&alphar,&alphai,&beta,&q,&ldq,&z,&ldz,&m,&pl,&pr,&dif,&work,&lwork,&iwork,&liwork,&info)
```

Comment:
The `&dif` here appears to be wrong.

Reply:
when `lquery=-1`, `dif` is not referenced so it shouldn't matter what it is.

Reply:
It produces errors because you pass a `double **` instead of a `double *`.

Reply:
Which machine are you compiling with? Because that should be caught  with the CI if there is a pointer problem.

Reply:
Actually, the same holds for all of the other arrays (b, alpha, beta, q, z, etc.) Why are you taking a reference of a pointer for all of them? That's just wrong.

I'm compiling this on my own machine with gcc 9.3.0. Incompatible pointer types are set to be warnings, which is why it compiles on the CI, but that doesn't make it less wrong.

Reply:
You should not get only for this but actually thousands of it because that's what all of the wrappers do. The internal call has pass by pointer and the dummy arguments are generated by f2py and passed via their pointers, if there is a double pointer issue all CI should catch it hence my question. Because I just tried it and I get no warnings

Reply:
Ah, I see now (because the arguments are not used). But then below it should read

```
    <ftype2> intent(hide) :: dif
```

instead of 

```
    <ftype2> intent(hide),dimension(2) :: dif
```

right?

Reply:
maybe try again after `git clean -xdf`?

Reply:
You might have missed my last comment since we posted at the same time. I think that's the culprit.

Reply:
Indeed I have. Thanks